cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1266 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+
+use cedar_policy_core::ast::{
+    BinaryOp, EntityUID, Expr, ExprKind, Literal, PolicyID, PolicySet, RequestType, UnaryOp, Var,
+};
+use cedar_policy_core::entities::err::EntitiesError;
+use cedar_policy_core::impl_diagnostic_from_source_loc_field;
+use cedar_policy_core::parser::Loc;
+use miette::Diagnostic;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+use crate::ValidationError;
+use crate::{
+    typecheck::{PolicyCheck, Typechecker},
+    types::{EntityRecordKind, Type},
+    ValidationMode, ValidatorSchema,
+};
+
+/// Data structure storing what data is needed
+/// based on the the [`RequestType`].
+/// For each request type, the [`EntityManifest`] stores
+/// a [`RootAccessTrie`] of data to retrieve.
+///
+/// `T` represents an optional type annotation on each
+/// node in the [`AccessTrie`].
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EntityManifest<T = ()>
+where
+    T: Clone,
+{
+    /// A map from request types to [`RootAccessTrie`]s.
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub per_action: HashMap<RequestType, RootAccessTrie<T>>,
+}
+
+/// A map of data fields to [`AccessTrie`]s.
+/// The keys to this map form the edges in the access trie,
+/// pointing to sub-tries.
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+pub type Fields<T> = HashMap<SmolStr, Box<AccessTrie<T>>>;
+
+/// The root of a data path or [`RootAccessTrie`].
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub enum EntityRoot {
+    /// Literal entity ids
+    Literal(EntityUID),
```

Comment:
I think this will end up exporting `cedar_policy_core::ast::EntityUID` from `cedar_policy`. Not a blocking issue for an experimental feature, but we'll need to introduce some more wrapper type in `cedar_policy` if we want callers to be able to use this directly as a `cedar_policy::EntityUID`. Same comment for `Var` variant.

Reply:
Do you mean that this should use the existing wrapper `EntityUid`?
That's a little unfortunate since I can't import it from the validator crate.

Reply:
Right, Callers should be able to use it as they would the existing `EntityUid` wrapper. It does get a annoying since you'll end up wanting a wrapper for most of your structs, but we don't want consumers of `cedar-policy` to ever import `cedar-policy-core`

Reply:
The other option would be using the opaque inner struct pattern we use for error messages if users don't need to look inside the `Literal`.

Reply:
That makes sense. I think people should be able to get the literal, so for now I'll add a TODO to implement wrappers for everything in `api.rs`